File: leetcode_python/206.py
# 206. 反转链表
# https://leetcode-cn.com/problems/reverse-linked-list/
from typing import List
from collections import defaultdict
from collections import deque
from math import inf
import bisect
import heapq
'''
@File : 206.py
@Time : 2021/12/21 20:29:04
@Author : YuMin Zhang
@State : Indepeedent | Half-Depedent | Depedent 
@Thinking :
@Tag : Easy | Medium | Hard
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        if not head:
            return None
        pre = None
        curr = head
        while curr:
            tmp = curr.next
            curr.next = pre
            pre = curr
            curr = tmp
        return pre

File: leetcode_c_plus_plus/206.cpp
// https://leetcode-cn.com/problems/reverse-linked-list/
// 206. 反转链表
#include<map>
#include<set>
#include<vector>
#include<cmath>
#include<numeric>
#include<unordered_map>
using namespace std;
#define INT_MAX 2147483647
#define INT_MIN (-INT_MAX - 1)
/**
@File : 206.cpp
@Time : 2021/12/30 20:21:33
@Author : YuMin Zhang
@State : Indepeedent
@Thinking :
@Tag : Easy
**/
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode * pre = nullptr;
        ListNode * curr = head;
        ListNode * tmp = nullptr;
        while(head) {
            tmp = head->next;
            head->next = pre;
            pre = head;
            head = tmp;
        }
        return pre;
    }
};


class Solution {

public:
    ListNode * ans;
    ListNode* reverseList(ListNode* head) {
        if(head == nullptr) return nullptr;
        recursive(head);
        return ans;
    }

    ListNode * recursive(ListNode * head) {
        if(head->next == nullptr) {
            ans = head;
            return head;
        }
        ListNode * node = recursive(head->next);
        node->next = head;
        head->next = nullptr;
        return head;
    }
};